google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/StatementParser.java

```diff
@@ -0,0 +1,379 @@
+    } else if (isQuery(sql)) {
+      return ParsedStatement.query(statement, sql);
+    } else if (isUpdateStatement(sql)) {
+      return ParsedStatement.update(statement, sql);
+    } else if (isDdlStatement(sql)) {
+      return ParsedStatement.ddl(statement, sql);
+    }
+    return ParsedStatement.unknown(statement, sql);
+  }
+
+  /**
+   * Parses the given statement as a client-side statement. Client-side statements are statements
+   * that are never sent to Cloud Spanner, but that are interpreted by the Connection API and then
+   * translated into some action, such as for example starting a transaction or getting the last
+   * commit timestamp.
+   *
+   * @param sql The statement to try to parse as a client-side statement (without any comments).
+   * @return a valid {@link ClientSideStatement} or null if the statement is not a client-side
+   *     statement.
+   */
+  @VisibleForTesting
+  ClientSideStatementImpl parseClientSideStatement(String sql) {
+    for (ClientSideStatementImpl css : statements) {
+      if (css.matches(sql)) {
+        return css;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Checks whether the given statement is (probably) a DDL statement. The method does not check the
+   * validity of the statement, only if it is a DDL statement based on the first word in the
+   * statement.
+   *
+   * @param sql The statement to check (without any comments).
+   * @return <code>true</code> if the statement is a DDL statement (i.e. starts with 'CREATE',
+   *     'ALTER' or 'DROP').
+   */
+  boolean isDdlStatement(String sql) {
+    return statementStartsWith(sql, ddlStatements);
+  }
+
+  /**
+   * Checks whether the given statement is (probably) a SELECT query. The method does not check the
+   * validity of the statement, only if it is a SELECT statement based on the first word in the
+   * statement.
+   *
+   * @param sql The statement to check (without any comments).
+   * @return <code>true</code> if the statement is a SELECT statement (i.e. starts with 'SELECT').
+   */
+  boolean isQuery(String sql) {
+    return statementStartsWith(sql, selectStatements);
+  }
+
+  /**
+   * Checks whether the given statement is (probably) an update statement. The method does not check
+   * the validity of the statement, only if it is an update statement based on the first word in the
+   * statement.
+   *
+   * @param sql The statement to check (without any comments).
+   * @return <code>true</code> if the statement is a DML update statement (i.e. starts with
+   *     'INSERT', 'UPDATE' or 'DELETE').
+   */
+  boolean isUpdateStatement(String sql) {
+    return statementStartsWith(sql, dmlStatements);
+  }
+
+  private boolean statementStartsWith(String sql, List<String> checkStatements) {
+    Preconditions.checkNotNull(sql);
+    String[] tokens = sql.split("\\s+", 2);
+    if (tokens.length > 0) {
+      for (String check : checkStatements) {
+        if (tokens[0].equalsIgnoreCase(check)) {
```

Comment:
can this be simplified to `sql.startsWith(check)` (handling upper and lower case)?

Reply:
Not really (at least not with the goal of simplifying). The problems with that is that we need a regular expression check for all types of spaces (space, tab, enter, ...), as any type of space character is allowed as a separator between for example CREATE and TABLE. The nice thing with the `String#split(String, int)` method is that it stops searching after finding the requested matches. `String.startsWith(String)` does not support regular expressions.